### PR TITLE
Native Histograms: automatically reduce resolution rather than fail scrape

### PR DESCRIPTION
**Pull request by fatsheep9146**
related #12864

when `native_histogram_bucket_limit` is hitted, reduce the resolution until it is beyond the limit.

### CODE REVIEW

File: scrape/target_test.go
Comment (bboreham) [inline scrape/target_test.go:524]: Should this test that the resolution was reduced?

Comment (fatsheep9146) [reply scrape/target_test.go:524]: Yes, since in this test, original bucket count is 6, the smallest possible buckets count we can get is 4, we can't reduce more to get smaller buckets count any more. So I add this test case.

I will try to add more comment or change the test case's name to describe the meaning of the test.

Comment (krajorama) [reply scrape/target_test.go:524]: I'd add `expectBucketCount` to verify the bucket count after the limit check. This would make it more obvious the semantics that the limit check also alters the buckets.

File: scrape/target.go
Comment (krajorama) [inline scrape/target.go:379]: Same as above, please rename `lastSum` to `lastCount`

**Review by krajorama — CHANGES_REQUESTED** [review]

File: scrape/target.go
Comment (krajorama) [inline scrape/target.go:372]: Isn't it possible that we have two buckets so far apart that one reduction doesn't merge them, but two reductions  does? In my PR in Mimir I actually do reductions until I reach the minimal schema (-4) instead of checking whether the number of buckets has decreased like here.

File: scrape/target_test.go
Comment (fatsheep9146) [reply scrape/target_test.go]: In addition to `expectBucketCount`, I also add `expectSchema`  to check if the schema is reduced as expected.

File: scrape/target.go
Comment (fatsheep9146) [reply scrape/target.go]: In this line I want to express that, we will reduce the resolution until we can not merge more buckets by reducing resolutions any more.

We can take the histogram in unit test `TestBucketLimitAppender` as example.

```
	example := histogram.Histogram{
		Schema:        0,
		Count:         21,
		Sum:           33,
		ZeroThreshold: 0.001,
		ZeroCount:     3,
		PositiveSpans: []histogram.Span{
			{Offset: 0, Length: 3},
		},
		PositiveBuckets: []int64{3, 0, 0},
		NegativeSpans: []histogram.Span{
			{Offset: 0, Length: 3},
		},
		NegativeBuckets: []int64{3, 0, 0},
	}
``` 

When schema is `0`, the positive and negative spans will be `{Offset: 0, Length: 3}` and `{Offset: 0, Length: 3}`, bucket counts is **6**.

When scheme is reduced to `-1`, the positive and negative spans will be `{Offset: 0, Length: 2}` and `{Offset: 0, Length: 2}`, bucket counts is **4**.

When schema is reduced to `-2`, the positive and negative spans will be `{Offset: 0, Length: 2}` and `{Offset: 0, Length: 2}`, bucket counts is still **4**.

So in this case, we can know that the bucket counts will not decrease even we reduce the resolution to -3, -4. So we can just break the loop.

@krajorama What do you think is the best way to break the loop of reducing the resolution?

Comment (krajorama) [reply scrape/target.go]: Hi, 

for example:
```
	bigGap := histogram.Histogram{
		Schema:        0,
		Count:         21,
		Sum:           33,
		ZeroThreshold: 0.001,
		ZeroCount:     3,
		PositiveSpans: []histogram.Span{
			{Offset: 1, Length: 1}, // in (1, 2]
			{Offset: 2, Length: 1}, // in (8, 16]
		},
		PositiveBuckets: []int64{1, 0}, // 1, 1
	}
```

Reducing the above histogram by one resolution does not reduce the number of buckets as the observations will be in
```
(1,4] and (4, 16]
```

However in the next step you get
```
(1, 16]
```

So I'd expect:
```
		{
			h: bigGap,
			limit: 1,
			expectError: false,
			expectBucketCount: 1,
			expectSchema: -2,
		},
```

Basically you have to continue until there are too many buckets and you haven't reached minimum schema.

e.g.:
```
	if h != nil {
		for len(h.PositiveBuckets)+len(h.NegativeBuckets) > app.limit {
			if h.Schema == -4 {
				return 0, errBucketLimit
			}
			h = h.ReduceResolution(h.Schema - 1)
		}
	}
	if fh != nil {
		for len(fh.PositiveBuckets)+len(fh.NegativeBuckets) > app.limit {
			if fh.Schema == -4 {
				return 0, errBucketLimit
			}
			fh = fh.ReduceResolution(fh.Schema - 1)
		}
	}
```

Comment (fatsheep9146) [reply scrape/target.go]: @krajorama I already change as you suggested.

File: scrape/target.go
Comment (krajorama) [inline scrape/target.go:370]: nit: would be nice to have an exported constant for this, I had to redefine in [Mimir](https://github.com/grafana/mimir/blob/6eb5f34818b32c91d113101e7a6cb16d4842f3b0/pkg/mimirpb/custom.go#L63) as well. A follow PR is ok I think.

File: scrape/target_test.go
Comment (krajorama) [inline scrape/target_test.go:530]: please add my bigGap test case to cover that

**Review by krajorama — APPROVED** [review]
Approved with comments. I'm not maintainer, so we need also @beorn7 or somebody to approve and merge.

File: scrape/target.go
Comment (fatsheep9146) [reply scrape/target.go:370]: Ok, but I'm not sure where to define, in prometheus or client_golang? @beorn7 do you have any suggestions?

File: scrape/target_test.go
Comment (fatsheep9146) [reply scrape/target_test.go:530]: Done

File: scrape/target.go
Comment (beorn7) [reply scrape/target.go:370]: I'm not so sure we should do it cross-repo. It's currently "part of the (not yet written) standard", but it might also get updated if need be. And then it might very well be that Prometheus itself supports a wider range of schemas than client_golang (e.g. if it is only needed to read OTel properly, while client_golang isn't concerned with reading OTel).

As a first step, there could be a constant in prometheus/prometheus/model/histogram.

But as said, doesn't have to be this PR.

**Review by beorn7 — APPROVED** [review]

**Pull request merged**
